content/blog/new-per-subject-discard-policy.md

```diff
@@ -0,0 +1,217 @@
+- Kafka also has a 'connection scoped window' deduplication mechanism built-in referred to as producer idempotence. When enabled the brokers remember all the request IDs received from a client connection ID and detects retries, which means that the producer can only guarantee idempotence for messages sent within a single session.
+
+As you can probably guess both methods have some clear limitations back in the real world of production systems:
+
+- The limitation of Kafka's producer idempotence being that you can never guarantee that the application publishing will never crash or be killed before receiving the publication's acknowledgement and therefore publish a duplicate when it restarts. And even then you still need to persist the positive acknowledgements that you have received to some kind of external datastore like Redis, so more than likely you are going to need to write and deploy something using Kafka Streams.
+- The limitation of JetStream's stream deduplication window being that one can never tell ahead of time how much time it will actually take before a failing client process can be restarted and able to successfully process the messages in the stream.
+
+## What is the feature itself?
+
+Now this finally brings us to the part where I describe the feature itself 🎉
+
+As the title of this blog implies, what this new feature of JetStream enables is the ability to have exactly-once publication through deduplication by BYOID over the entire stream, regardless of time between fault and restart, of client disconnections, or of the client or server processes getting restarted.
+
+So, how do you actually use this new feature, and how does it work? Bear with me again as I first go through some prerequisite knowledge about stream limits and discard policies.
+
+### Stream limits
+
+JetStream allows you to set basic limits for a stream: the maximum number of messages in the stream, the maximum number of bytes, and you can specify the max amount of time you want the stream to keep messages (i.e. a TTL). On the other hand, Kafka and other distributed log systems allow you to adjust when log compaction kicks in to prune the tail end of the stream data.
+
+### Discard policies
+
+Because it is a data store, JetStream also allows you to specify a 'discard policy': what happens when a limit is breached because of a new publication. There are two discard policies:
+
+- Discard `old` means that the stream will discard the oldest message in the stream in order to accept the new message being published while keeping the stream at the limit.
+
+This is probably the most common use-case: you want the stream to be a rolling window of data and yet control the size of the stream, new publications silently push out the oldest message in the stream. It is the only kind of discard policy that append only log streaming system with log compaction systems support, as they are meant to always accept new messages or block until they can.
```

Comment:
That JS has the concept of 'old' and 'new' discard policies, while log systems don't have the concept of a discard policy because they only can support the equivalent of the 'old' discard policy (because they always either accept the message to the stream of block until it can be accepted)